From c85010272f6e393aa4f90d6c53a11a9f6aaf89f0 Mon Sep 17 00:00:00 2001
From: Natanael Copa <ncopa@alpinelinux.org>
Date: Mon, 8 May 2023 20:17:06 +0200
Subject: [PATCH 4/5] alpine-config: add apk_cache, apk_repositories and
 packages

So we can configure apk cache, repositories and install packages from
user-data.
---
 lib/tiny-cloud/user-data/alpine-config | 67 +++++++++++++++++-
 tests/tiny-cloud-alpine.test           | 95 +++++++++++++++++++++++++-
 2 files changed, 160 insertions(+), 2 deletions(-)

diff --git a/lib/tiny-cloud/user-data/alpine-config b/lib/tiny-cloud/user-data/alpine-config
index 4dd1134..dfff109 100644
--- a/lib/tiny-cloud/user-data/alpine-config
+++ b/lib/tiny-cloud/user-data/alpine-config
@@ -1,7 +1,9 @@
 # Script UserData Functions
 # vim:set ts=4 et ft=sh:
 
-INIT_ACTIONS_MAIN="$(insert_after set_hostname userdata_ntp $INIT_ACTIONS_MAIN)"
+INIT_ACTIONS_MAIN="$(insert_after set_hostname \
+    "userdata_ntp userdata_apk_cache userdata_apk_repositories userdata_packages" \
+    $INIT_ACTIONS_MAIN)"
 
 init__userdata_ntp() {
     local ntp_enabled="$(imds user-data/ntp/enabled)"
@@ -31,3 +33,66 @@ init__userdata_ntp() {
         $MOCK rc-service "$svc" start
     fi
 }
+
+init__userdata_apk_cache() {
+    local cache="$(imds user-data/apk/cache)"
+    if [ -z "$cache" ]; then
+        return
+    fi
+    mkdir -p "$ROOT/$cache"
+    # make link relative
+    case "$cache" in
+        /*) cache="../..$cache";;
+    esac
+    mkdir -p "$ROOT"/etc/apk
+    ln -sf "$cache" "$ROOT"/etc/apk/cache
+}
+
+init__userdata_apk_cache() {
+    local cache="$(imds user-data/apk/cache)"
+    if [ -z "$cache" ]; then
+        return
+    fi
+    mkdir -p "$ROOT/$cache"
+    # make link relative
+    case "$cache" in
+        /*) cache="../..$cache";;
+    esac
+    mkdir -p "$ROOT"/etc/apk
+    ln -sf "$cache" "$ROOT"/etc/apk/cache
+}
+
+init__userdata_apk_repositories() {
+    local repositories="$(imds user-data/apk/repositories)"
+    mkdir -p "$ROOT"/etc/apk
+    for r in $repositories; do
+        local baseurl="$(imds user-data/apk/repositories/$r/base_url)"
+        local repos="$(imds user-data/apk/repositories/$r/repos)"
+        local version="$(imds user-data/apk/repositories/$r/version)"
+        if [ -z "$version" ]; then
+            local version_id=$( . "$ROOT"/etc/os-release 2>/dev/null && echo "$VERSION_ID")
+            case "$version_id" in
+                edge*|*_alpha*) version="edge";;
+                [0-9]*.[0-9]*.[0-9]*) version="v${version_id%.*}";;
+            esac
+        fi
+        if [ -n "$version" ] && [ "$version" != "." ] && [ "$version" != "/" ]; then
+            baseurl="${baseurl%/}/$version"
+        fi
+        for repo in $repos; do
+            local uri="${baseurl%/}/$(imds user-data/apk/repositories/$r/repos/$repo)"
+            add_once "$ROOT"/etc/apk/repositories "$uri"
+        done
+    done
+}
+
+init__userdata_packages() {
+    local packages="$(imds user-data/packages)"
+    local pkgs=
+    for i in $packages; do
+        pkgs="$pkgs $(imds user-data/packages/$i)"
+    done
+    if [ -n "$pkgs" ]; then
+        $MOCK apk add $pkgs
+    fi
+}
diff --git a/tests/tiny-cloud-alpine.test b/tests/tiny-cloud-alpine.test
index 43a0161..60e5c4a 100755
--- a/tests/tiny-cloud-alpine.test
+++ b/tests/tiny-cloud-alpine.test
@@ -12,7 +12,13 @@ init_tests \
 	set_network_config_auto \
 	userdata_ntp \
 	userdata_ntp_busybox \
-	userdata_ntp_openntpd
+	userdata_ntp_openntpd \
+	userdata_apk_cache \
+	userdata_apk_repositories \
+	userdata_apk_repositories_version \
+	userdata_apk_repositories_version_auto_edge \
+	userdata_packages
+
 
 set_network_config_network_interfaces_body() {
 	fake_metadata_nocloud <<-EOF
@@ -115,3 +121,90 @@ userdata_ntp_openntpd_body() {
 		-o match:"rc-service .*openntpd" \
 		tiny-cloud main
 }
+
+userdata_apk_cache_body() {
+	fake_userdata_nocloud <<-EOF
+		#alpine-config
+		apk:
+		  cache: /var/cache/apk
+	EOF
+	atf_check -e ignore -o ignore tiny-cloud net
+	atf_check \
+		-e match:"userdata_apk_cache .*DONE" \
+		-o ignore \
+		tiny-cloud main
+	atf_check -o match:"$PWD/var/cache/apk" readlink -f etc/apk/cache
+}
+
+userdata_apk_repositories_body() {
+	fake_userdata_nocloud <<-EOF
+		#alpine-config
+		apk:
+		  repositories:
+		    - base_url: /srv/packages
+		      repos: [ "main", "community" ]
+	EOF
+	atf_check -e ignore -o ignore tiny-cloud net
+	atf_check \
+		-e match:"userdata_apk_repositories .*DONE" \
+		-o ignore \
+		tiny-cloud main
+	atf_check -o match:"^/srv/packages/main$" \
+		-o match:"^/srv/packages/community$" \
+		cat etc/apk/repositories
+}
+
+userdata_apk_repositories_version_body() {
+	fake_userdata_nocloud <<-EOF
+		#alpine-config
+		apk:
+		  repositories:
+		    - base_url: https://cdn.alpinelinux.org/
+		      version: edge
+		      repos: [ "main", "community" ]
+	EOF
+	atf_check -e ignore -o ignore tiny-cloud net
+	atf_check \
+		-e match:"userdata_apk_repositories .*DONE" \
+		-o ignore \
+		tiny-cloud main
+	atf_check -o match:"^https://cdn.alpinelinux.org/edge/main$" \
+		-o match:"^https://cdn.alpinelinux.org/edge/community$" \
+		cat etc/apk/repositories
+}
+
+userdata_apk_repositories_version_auto_edge_body() {
+	fake_userdata_nocloud <<-EOF
+		#alpine-config
+		apk:
+		  repositories:
+		    - base_url: https://cdn.alpinelinux.org/
+		      repos: [ "main", "community" ]
+	EOF
+	mkdir -p etc
+	echo "VERSION_ID=3.18_alpha20230329" > etc/os-release
+
+	atf_check -e ignore -o ignore tiny-cloud net
+	atf_check \
+		-e match:"userdata_apk_repositories .*DONE" \
+		-o ignore \
+		tiny-cloud main
+	atf_check -o match:"^https://cdn.alpinelinux.org/edge/main$" \
+		-o match:"^https://cdn.alpinelinux.org/edge/community$" \
+		cat etc/apk/repositories
+}
+
+userdata_packages_body() {
+	fake_userdata_nocloud <<-EOF
+		#alpine-config
+		packages:
+		  - tmux
+		  - vim
+	EOF
+	atf_check -e ignore -o ignore tiny-cloud net
+	atf_check \
+		-e match:"userdata_packages .*DONE" \
+		-o match:"apk add .*tmux" \
+		-o match:"apk add .*vim" \
+		tiny-cloud main
+}
-- 
2.40.1

